src/orchestrator/orchestrator.rs

```diff
@@ -54,45 +56,71 @@ impl<const N_ASSETS: usize, const N_BYTES: usize> Orchestrator<N_ASSETS, N_BYTES
+        let entries_per_executor = self.entry_csvs.len() / executor_count;
 
-        // Declare channels for internal communication
-        // One of the channels receives parsed data from entry parser to distribute tasks to executor
-        // while the other channel is used the executors that send a result of the tasks.
         let mut executors = Vec::new();
         let mut result_collectors = Vec::new();
 
-        let actual_number_of_workers = min(number_executor, self.entry_csvs.len());
-        for i in 0..actual_number_of_workers {
-            let (entries_tx, mut entries_rx) = mpsc::channel(32);
-            let (tree_tx, tree_rx) = mpsc::channel(32);
-            let executor = self.executor_spawner.spawn_executor().await;
-            result_collectors.push((i, tree_rx));
+        let channel_size = std::env::var("CHANNEL_SIZE")
+            .unwrap_or_default()
+            .parse::<usize>()
+            .unwrap_or(32);
 
-            // 2. Executor
+        let cancel_token = CancellationToken::new();
+        let actual_number_of_workers = min(executor_count, self.entry_csvs.len());
+        for i in 0..actual_number_of_workers {
+            // Declare channels for communication
             //
-            // Spawn executors that process entries with Worker.
+            // There are three channels are used inthis method.
             //
-            // - Receives 'entries' from [entries_rx] channel.
-            // - Processes 'entries' to build a tree (done by worker).
-            // - Sends the resulting 'tree' back via [tree_tx] channel.
+            // - A `entries_tx` receives parsed data from the entry parser to distribute tasks to executors.
+            // - A `tree_tx` channel is used by the executors to send the results of the tasks.
+            // - A `stop` channel is used to send a stop signal to all executors in case of an error,
+            //   allowing for a coordinated shutdown or cancellation of tasks.
             //
+            let (entries_tx, mut entries_rx) = mpsc::channel(channel_size);
+            let (tree_tx, tree_rx) = mpsc::channel(channel_size);
+
+            let executor = self.executor_spawner.spawn_executor().await;
+            result_collectors.push((i, tree_rx));
+
+            let cloned_cancel_token = cancel_token.clone();
             executors.push(tokio::spawn(async move {
-                while let Some(task) = entries_rx.recv().await {
-                    let processed_task = executor
-                        .generate_tree::<N_ASSETS, N_BYTES>(task)
-                        .await
-                        .unwrap();
-                    if tree_tx.send(processed_task).await.is_err() {
-                        break;
-                    }
-                }
+                        loop {
+                            tokio::select! {
+                                entries_data = entries_rx.recv() => {
```

Comment:
If this `recv` doesn't have an await, does it mean that the loop would try to create a lot of instances very quickly?

Reply:
> If this `recv` doesn't have an await, does it mean that the loop would try to create a lot of instances very quickly?

No, the `recv` function includes an await internally. FYI, This is part of a pattern structure `<pattern> = <async expression> => <handler>` used within the `select` macro. 

For a more detailed explanation, please refer to https://tokio.rs/tokio/tutorial/select#syntax